README.org
* Matlab-emacs Project:

MATLAB and Emacs integration

** INSTALL

  Add the following to your .emacs file:

  ;; Replace path below to be where your matlab.el file is.
  (add-to-list 'load-path "~/path/to/matlab-emacs")
  (load-library "matlab-load")

  ;; Enable CEDET feature support for MATLAB code. (Optional)
  ;; (matlab-cedet-setup)

** MELPA

GNU emacs (sorry Xemacs folks) users (version >=24) can use the
package as provided by MELPA.
Set 
(add-to-list 'package-archives
             '("melpa" . "http://melpa.org/packages/") t)

And then run 
M-x
package-list-packages



** Releases

   The most recent version of matlab.el is always in GIT.

   You can get an old downloader for matlab-emacs from MATLAB File Exchange.

   Category:
   http://www.mathworks.com/matlabcentral/fileexchange/loadCategory.do?objectId=19&objectType=Category
   This File
   http://www.mathworks.com/matlabcentral/fileexchange/loadFile.do?objectId=104&objectType=file

   Older versions of matlab.el can be found as matlab.el.1.10.1
   for emacsen that do not have the latest additional utilities such
   as tempo and derived.

   matlab.el and mlint.el has shipped on MATLAB CDs starting with 6.5
   (R13).  Updates after a release of MATLAB will be made to MATLAB
   Central and announced on the mailing list.

   matlab.el and mlint.el will no longer ship on MATLAB CDs starting
   with MATLAB release R2008b.  If you are using a version of MATLAB
   from before R2008a, you will want to use the version of mlint from
   that CD.

** GIT

   matlab-emacs is *now* available via GIT from SourceForge. The CVS
   repository is still at SourceForge but it is *no* longer updated.

   Homepage:
   http://matlab-emacs.sf.net

   Project Page:
   http://sourceforge.net/projects/matlab-emacs

   GIT Repository:

   https://sourceforge.net/p/matlab-emacs/src/ci/master/tree/

   You need a git client on your machine. Then just run


   git clone git://git.code.sf.net/p/matlab-emacs/src matlab-emacs-src

   If you do not have a GIT client on your machine, you can use the
   MATLAB script dl_emacs_support.m to download a fresh copy of the
   matlab.el sources.
   https://sourceforge.net/p/matlab-emacs/src/ci/master/tree/dl_emacs_support.m


** Dependencies

   MATLAB-Emacs can use the CEDET suite for the following items:
	- Modifying the build system (Makefiles, etc)
	- mlint (uses EIEIO object system)
	- parsing/completion (uses semantic parsing system)
        - some template insertion features

   As of Emacs 23.2, CEDET is included in Emacs, and nothing extra
   is needed to satisfy that dependency.

   For older versions of Emacs:

   See http://cedet.sf.net for downloading CEDET.

   If you only want to use matlab.el for editing code or running the
   MATLAB shell, you DO NOT need to install CEDET.

   If you do want to use CEDET, add the following to your .emacs file:

   (matlab-cedet-setup)

** New Mailing List

   A new matlab-emacs-discuss mailing lisit has been created at
   SourceForge.  Use this list to get updates on new releases and
   discuss topics relivant to the matlab-emacs project.

   http://lists.sourceforge.net/mailman/listinfo/matlab-emacs-discuss

** Old Mailing List

   The original mailing list where beta versions of matlab.el were
   posted, and where comments, questions, bug reports, and answers to
   questions could be sent.  If you were subscribed to this list,
   please unsubscribe, and subscribe to the new list above.

   to unsubscribe, send email with the body of: "unsubscribe matlab-emacs"

** Known problems

*** Unable to open .m files in matlab with Emacs

From: Xu He <xuhe01@gmail.com> wrote:

Dear all,

Following the install guide, I setup the matlab-emacs on my ubuntu 12.04
system, and I can edit and run the m file with emacs.

But I can not open nor edit m file with matlab. When I open the file in the matlab window,
nothing happens.

I tried to eliminate the lines about matlab-emacs in ~/.emacs file, but the
problem can not be solved.

I need to edit and run the file in matlab. Thanks for any help.

Best regards,
Xu He

*** Solution 

From: Hunter McClelland [mailto:hunter.mcclelland@gmail.com]
Till we discovered that Matlab-emacs changes the "Editor" matlab uses.
Go to Preferences->Editor/debugger and change the Editor back to
"MATLAB Editor".


ChangeLog
2016-03-17  Uwe Brauer  <oub@mat.ucm.es>

	* toolbox/dbhotlink.m (dbhotlink): new file

2016-03-17  Uwe Brauer  <oub@mat.ucm.es>

	* matlab.el (matlab-shell-mode): Applied patch from  Odd Andersen
	<odd.andersen@gmail.com>. 

2016-03-16  Uwe Brauer  <oub@mat.ucm.es>

	* dl_emacs_support.m (getfiles): update the url and README-->README.org

2016-02-09  Uwe Brauer  <oub@mat.ucm.es> 

	* matlab.el (matlab-mode-version): Change version number to
	3.3.3

	* matlab-publish.el (matlab-select-publish-form): Cleanup file

	* matlab.el (matlab-change-current-directory): Applied patch,
	from an unknown source found in http://sourceforge.net/p/matlab-emacs/patches/2/
	Purpose: add the possibility to switch the matlab directory to the
	current one.

	* matlab.el (matlab-shell): Applied patch
	from http://sourceforge.net/p/matlab-emacs/patches/2/
	which results in
	(define-key km (kbd "TAB") 'matlab-shell-tab)

2016-02-09  Uwe Brauer  <oub@mat.ucm.es>

	Imported from CVS to git.

2014-11-05  zappo  <zappo@eludlam-deb6-64>

	* Makefile: (VERSION) Updated.
	(all, tags, dist): Add toolbox
	(toolbox): New rule

	* Project.ede (:version): Updated

	* templates/Makefile (VERSION): Update to minor rev.

	* toolbox/Makefile: Makefile for toolbox.

	* toolbox/Project.ede: Project file for toolbox.

	* INSTALL (tlc.el): Fix typo.

	* matlab.el (matlab-shell-completion-list):
	remove custom call to MCR with call to emacsdocomplete.m in the matlab-emacs toolbox directlry.

	* toolbox/emacsdocomplete.m:
	Command for getting completions from MATLAB for matlab-shell.

2014-11-05  zappo  <zappo@eludlam-deb6-64>

	* matlab.el (matlab-shell-completion-list):
	remove custom call to MCR with call to emacsdocomplete.m in the matlab-emacs toolbox directlry.

	* toolbox/emacsdocomplete.m:
	Command for getting completions from MATLAB for matlab-shell.

2014-09-24  zappo  <zappo@eludlam-deb6-64>

	* .cvsignore: First checkin.

2014-09-23  zappo  <zappo@eludlam-deb6-64>

	* matlab.el (matlab-shell-font-lock-keywords):
	Tweak to support some changes in error output.
	(matlab-comment): For comment on same line.  If we run out of space (fill-column)
	pull the comment closer to the code.
	(matlab-auto-fill): When filling a string, improve checking of when to add brackets
	around the string.
	(matlab-shell-mode): Disable html rendering trickery.
	(gud-matlab-error-regexp): Add a new variant (off by default) as reminder to
	go back and get the current ones to handle newer MATLAB error format.

2014-03-07  zappo  <zappo@eludlam-deb6-64>

	* mlint.el (mlint-minor-mode): Remove make-local-hook for Emacs 24.

2013-08-29  zappo  <zappo@eludlam-deb6-64>

	* semantic-matlab.el (semantic/dep): New dependency (Emacs 24)
	(semantic-matlab-parse-oldstyle-class): Add 'method' local var to tidy
	compile warnings.

2013-08-26  zappo  <zappo@eludlam-deb6-64>

	* matlab.el (matlab-mode-version): Bump very-minor versin.
	(matlab-ltype-comm): Add support for block comment detection.
	(matlab-ltype-block-comm): New fcn.
	(matlab-shell): Add TAB binding for TAB for when (tab) doesn't work.
	Add support for setting WINDOWID environment variable so that MATLAB
	will bring Emacs forward when typing in a figure.

2013-04-02  zappo  <zappo@eludlam-deb6-64>

	* README: Update to discuss CEDET that comes with Emacs.

	* INSTALL: Update to note latest build notes.

	* Makefile: Regenerated
	Added CEDET_PATH variable if someone wants to download the latest from bzr.

	* Project.ede (:version): Updated
	(lisp): Add linemark.el

	* matlab-load.el: Regenerated.

	* dl_emacs_support.m (coreFiles): Add linemark.el to download list.

	* linemark.el:
	Add linemark to matlab-emacs dist so that mlint will work
	with the version of CEDET integrated with Emacs.

	* cedet-matlab.el (matlab-cedet-setup):
	Update srecode require to work w/ CEDET
	integrated w/ Emacs.

	* semantic-matlab.el (misc require statements):
	Updated to work w/ CEDET integrated w/ Emacs.

	* semanticdb-matlab.el (misc require statements):
	Updated to work w/ CEDET integrated w/ Emacs.
	(semanticdb-matlab-scan-directories): Remove references to `working'.
	(semanticdb-matlab-cache-files): Remove references to `working'.

	* templates/Makefile (VERSION): Updated.

2012-02-23  zappo  <zappo@eludlam-deb6-64>

	* matlab.el (matlab-mode):
	Change page-delimeter to allow any whitespace or \n after a %%.

2011-10-05  zappo  <zappo@eludlam-deb6-64>

	* matlab.el (matlab-ltype-endfunction-comm):
	If the endfunction comment occurs on
	a line, but the next bit of code isn't part of an enclosing fcn, then no.
	(matlab-shell-mode): Patch from Eli Merriam;
	make comint-input-filter-functions buffer local.
	Add matlab-shell-render-html-txt-format hook to comint.
	(matlab-txt-format-beg, matlab-txt-format-end): New
	(matlab-shell-render-html-txt-format): New.
	(matlab-shell-last-anchor-as-frame): New.
	(matlab-shell-render-errors-as-anchor): Use last anchor var above.
	Note: Not used for any logic yet. ;(
	(matlab-shell-previous-matching-input-from-input): patch; Mark Histed
	Force cursor to eol.
	(matlab-shell-run-region): Force displayed buffer to recycled visible buffers.

2011-08-10  davenar  <davenar@eludlam-deb6-64>

	* matlab.el (matlab-shell-mode):
	Add `comint-postoutput-scroll-to-bottom' to
	`comint-output-filter-functions' (suggested by Mark Histed).

2011-08-09  davenar  <davenar@eludlam-deb6-64>

	* matlab.el (matlab-shell-run-region):
	Fix last commit and correctly use
	`buffer-substring' again.

	* matlab.el (matlab-shell-run-region):
	When removing comments, do not touch matlab
	strings or we might remove format specifiers.

2011-07-17  davenar  <davenar@eludlam-deb6-64>

	* matlab.el (matlab-shell-run-region):
	Remove all comments before running a
	region, otherwise `matlab-shell-run-cell' will break.

	* matlab.el (matlab-shell-run-region):
	Remove comment lines.  If NOSHOW, also
	remove continuations.

	* matlab.el (matlab-shell-run-region):
	New optional argument NOSHOW to replace

matlab-publish.el
;;; matlab-publish.el --- Utilities for editing MATLAB files for publishing

;; Copyright (C) 2009 Uwe Brauer

;; Author: Uwe Brauer oub@mat.ucm.es
;; Maintainer: Uwe Brauer oub@mat.ucm.es
;; Created: 25 Feb 2009
;; Version: 1.0
;; Keywords:

 
;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 1, or (at your option)
;; any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; A copy of the GNU General Public License can be obtained from this
;; program's author (send electronic mail to oub@mat.ucm.es) or from
;; the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
;; 02139, USA.

;; LCD Archive Entry:
;; my-pub|Uwe Brauer|oub@mat.ucm.es
;; |
;; |$Date$|$Revision$|~/packages/my-pub.el

;;; Commentary:

;;{{{Change log:

;; $Id$
;; $Log$
;; Revision 1.1  2009-07-06 19:59:10  zappo
;; Utilities for editing MATLAB files for publishing
;;
;; Revision 1.9  2009/03/08 09:38:31  oub
;; 	* matlab-publish.el (matlab-select-environment): change function
;; 	slightly
;; 	(matlab-insert-section): new function
;;
;; Revision 1.8  2009/03/07 13:54:57  oub
;; 	  (matlab-write-region-novisit): New function
;; 	  (matlab-publish-region-old): new function
;; 	  (matlab-publish-region): New function
;; 	  (matlab-publish-region-latex): new function
;; 	  (matlab-publish-region-html): new function
;;
;; Revision 1.7  2009/03/07 13:26:55  oub
;; 	(matlab-show-matlab-shell): New variable
;; 	(matlab-publish-file-latex): add
;; 	  (if matlab-show-matlab-shell
;; 	  (matlab-show-matlab-shell-buffer)))
;;
;; Revision 1.6  2009/03/07 11:00:04  oub
;; 	* matlab-publish.el (matlab-boldify): Modify the function.
;;
;; Revision 1.5  2009/03/06 14:24:47  oub
;; 	* matlab-publish.el (matlab-temp-region-file): new variable
;; 	(matlab-write-region): New function
;;
;; Revision 1.4  2009/02/26 16:18:23  oub
;; modifiy the insert functions in order that the jump to the text where
;; new stuff has to be inserted.
;;
;; Revision 1.3  2009/02/26 16:06:08  oub
;; (defun matlab-select-environment (ch)
;;   (interactive "c1: title, 2: descr, 3: bold:, 4:mono, 5:pre, 6:equation, 7:list ")
;;   (setq ch (downcase ch))
;;   (call-interactively (cond ((eql ch ?1) #'matlab-insert-title)
;;                             ((eql ch ?2) #'matlab-insert-description-text)
;;                             ((eql ch ?3) #'matlab-insert-bold-text)
;;                             ((eql ch ?4) #'matlab-insert-monospaces-text)
;;                             ((eql ch ?5) #'matlab-insert-preformated-text)
;;                             ((eql ch ?6) #'matlab-insert-equation)
;;                             ((eql ch ?7) #'matlab-insert-bullet-list)
;;                             (t (error 'args-out-of-range '(1 2 3 5 6 7 ch))))))
;;
;; Revision 1.2  2009/02/26 11:24:17  oub
;; New functions
;; (defun matlab-insert-title ()
;;
;; (defun matlab-insert-description-text ()
;;
;;
;; (defun matlab-insert-bold-text ()
;;
;;
;; (defun matlab-insert-monospaces-text ()
;;
;;
;;
;; (defun matlab-insert-preformated-text ()
;;
;;
;; (defun matlab-insert-equation ()
;;
;;
;; (defun matlab-insert-bullet-list ()
;;
;; Revision 1.1  2009/02/26 11:15:02  oub
;; Initial revision
;;

;;}}}

;;; Code:


;; (defun my-publish-matlab ()
;;   (interactive)
;;   (let ((pub (file-name-nondirectory (buffer-file-name))))
;; 	(let ((arg (concat "publish('" pub "','latex')")))
;; 	  (matlab-shell-run-command arg))))

(require 'matlab)

;; Variables
(defvar matlab-temp-region-file "region.m"  ;Version-1.5
  "*Variable for the file which saves the region for publishing.")

(defvar matlab-show-matlab-shell t
"* t means the matlab buffer shell is shown when running matlab.")

;; Functions
(defun matlab-select-publish-form (ch)
"This function allows to publish the m file, either as in LaTeX or in
HTML format."
  (interactive "c1: LaTeX, 2: HTML ")
  (setq ch (downcase ch))
  (call-interactively (cond ((eql ch ?1) #'matlab-publish-file-latex)
                            ((eql ch ?2) #'matlab-publish-file-html)
                            (t (error 'args-out-of-range '(1 2 ch))))))

(defun matlab-publish-file-latex ()		;Version-1.7
"Publish a matlab file in the LaTeX format."
  (interactive)
  (let ((pub (file-name-nondirectory (buffer-file-name)))) 
	(matlab-shell-run-command (format "publish('%s','latex')" pub)))
  (if matlab-show-matlab-shell
	  (matlab-show-matlab-shell-buffer)))



(defun matlab-publish-file-html ()		;Version-1.7
  (interactive)
"Publish a matlab file in the html format."
  (let ((pub (file-name-nondirectory (buffer-file-name))))
 	(matlab-shell-run-command (format "publish('%s','html')" pub)))
  (if matlab-show-matlab-shell
	  (matlab-show-matlab-shell-buffer)))



(defun matlab-select-environment (ch) 	;Version-1.9
"This functions inserts structured text, which results for example 
in LaTeX mode in title, sections, description, boldified text,  unnumbered equations and bullet list."
  (interactive "c1: title, 2: section, 3:descrip, 4:boldify, 5:equation, 6:list ")
  (setq ch (downcase ch))
  (call-interactively (cond ((eql ch ?1) #'matlab-insert-title)
                            ((eql ch ?2) #'matlab-insert-section)
                            ((eql ch ?3) #'matlab-insert-description-text)
                            ((eql ch ?4) #'matlab-boldify)
                            ((eql ch ?5) #'matlab-insert-equation)
                            ((eql ch ?6) #'matlab-insert-bullet-list)
                            (t (error 'args-out-of-range '(1 2 3 5 6 ch))))))


(defun matlab-insert-title ()
  (interactive)
  (beginning-of-buffer nil)
  (insert "%% TITLE\n")
  (previous-line 1)
  (forward-char 3))

(defun matlab-insert-section ()		;Version-1.9
  (interactive)
  (insert "%% Section\n")
  (previous-line 1)
  (forward-char 3))


(defun matlab-insert-description-text ()
  (interactive)
  (insert "%%\n")
  (insert "% DESCRIPTIVE TEXT\n")
  (previous-line 1)
  (forward-char 2))



(defun matlab-boldify ()				;Version-1.6
"Insert either \%\%\\n \% \*BOLD TEXT\*\\n or, when mark is active,
surrounds region by * *."
  (interactive)
  (if (or (and (boundp 'zmacs-region-active-p) zmacs-region-active-p)
		  (and (boundp 'transient-mark-mode) transient-mark-mode mark-active))
	  (save-excursion
		(goto-char (point))
		(insert "*")
		(goto-char (mark))
		(insert "*"))
  (insert "\n%%\n")
  (insert "% *BOLD TEXT*\n")
	(backward-char 2)))


(defun matlab-insert-bold-text ()
  (interactive)
  (insert "%%\n")
  (insert "% *BOLD TEXT*\n")
  (previous-line 1)
  (forward-char 2))

(defun matlab-insert-monospaces-text ()
  (interactive)
  (insert "%%\n")
  (insert "% |MONOSPACED TEXT|\n")
  (previous-line 1)
  (forward-char 2))


(defun matlab-insert-preformated-text ()
  (interactive)
  (insert "%%\n")
  (insert "%\n") 
  (insert "%  PREFORMATTED\n")
  (insert "%  TEXT\n")
  (insert "% \n")
  (previous-line 3)
  (forward-char 3))

(defun matlab-insert-equation ()
  (interactive)
  (insert "%%\n")
  (insert "% \n")
  (insert "% $$e^{\pi i} + 1 = 0$$\n")
  (insert "% \n")
  (previous-line 2)
  (forward-char 4))


(defun matlab-insert-bullet-list ()
  (interactive)
  (insert "%%\n")
  (insert "% \n")
  (insert "% * ITEM1\n")
  (insert "% * ITEM2\n")
  (insert "% \n")
  (previous-line 3)
  (forward-char 4))




(defun matlab-write-region (start end)	;Version-1.5
  (interactive "r")
  (write-region start end (expand-file-name matlab-temp-region-file) nil nil nil nil)
  (find-file matlab-temp-region-file))


(defun matlab-write-region-novisit (start end)
  (interactive "r")
  (write-region start end (expand-file-name matlab-temp-region-file) nil nil nil nil))




(defun matlab-publish-region-old (ch)	;Version-1.8
  (interactive "c1: LaTeX, 2: HTML ")
  (setq ch (downcase ch))
  (matlab-write-region-novisit (region-beginning) (region-end))
  (call-interactively (cond ((eql ch ?1) #'matlab-publish-region-latex)
                            ((eql ch ?2) #'matlab-publish-region-html)
                            (t (error 'args-out-of-range '(1 2 ch))))))


(defun matlab-publish-region (start end ch)	 ;Version-1.8
  (interactive "r\nc1: LaTeX, 2: HTML ")
  (setq ch (downcase ch))
  (write-region start end (expand-file-name matlab-temp-region-file) nil nil nil nil)
  (call-interactively (cond ((eql ch ?1) #'matlab-publish-region-latex)
                            ((eql ch ?2) #'matlab-publish-region-html)
                            (t (error 'args-out-of-range '(1 2 ch))))))



(defun matlab-publish-region-latex ()	;Version-1.8
  (interactive)
	(matlab-shell-run-command (format "publish('%s','latex')" matlab-temp-region-file))
  (if matlab-show-matlab-shell
	  (matlab-show-matlab-shell-buffer)))



(defun matlab-publish-region-html ()	;Version-1.8
  (interactive)
	(matlab-shell-run-command (format "publish('%s','html')" matlab-temp-region-file))
  (if matlab-show-matlab-shell
	  (matlab-show-matlab-shell-buffer)))




(provide 'matlab-publish)

;;; MY-PUB.EL ends here

;; (defvar isp "hallo"
;; "*check")
;;
;; (when (string-match isp "hallo"))
;; (message "it works"))